# Customer Churn Prediction & Customer Segmentation

## 01 — Data Validation

### Business Context

The goal of this project is to help a telecom business identify customers who are likely to churn and understand different customer segments.

Before we perform EDA, feature engineering, or model training, we first need to make sure that the raw dataset is reliable.

This notebook is therefore focused on **Data Validation**.

We will answer:

- How large is the dataset?
- Are the expected columns present?
- Are there missing values?
- Are there blank values hidden inside text columns?
- Are there duplicate customers or duplicate rows?
- Are the data types appropriate?
- Are categorical values consistent?
- Are numerical values reasonable?
- Is the target variable valid?
- Are there any obvious data-quality or leakage risks?

> **Important:** We are not cleaning the dataset in this notebook.
> We are identifying problems first. Cleaning decisions will be made in the next stage.

## 1. Import the required libraries

We only import libraries that are needed for this validation stage.

At this point, we do not need machine-learning libraries because we are not training a model yet.

In [11]:
import pandas as pd
from pathlib import Path

## 2. Define the dataset path

The raw dataset is stored inside:

`data/raw/Customer-Churn.csv`

We keep the raw file unchanged. This is important because the raw dataset should always remain available as the original source of our analysis.

In [12]:
# Get the project root directory.
# The notebook is inside the "notebooks" folder,
# so the parent directory is the project root.

PROJECT_ROOT = Path.cwd().parent

# Build the complete path to the raw dataset.
DATA_PATH = PROJECT_ROOT / "data" / "raw" / "Customer-Churn.csv"

print("Dataset path:")
print(DATA_PATH)

Dataset path:
d:\Projects\Customer Churn Segmentation\data\raw\Customer-Churn.csv


## 3. Make sure the dataset exists

Before trying to read the CSV file, we check whether the expected file actually exists.

This gives us a clear error message instead of a confusing file-not-found error later.

In [13]:
if DATA_PATH.exists():
    print("Dataset file found.")
else:
    print("Dataset file was not found.")
    print("Expected location:", DATA_PATH)

Dataset file found.


## 4. Load the raw dataset

Now we can load the CSV file into a Pandas DataFrame.

We are intentionally using the raw dataset here without modifying any values.

In [14]:
df = pd.read_csv(DATA_PATH)

print("Dataset loaded successfully.")

Dataset loaded successfully.


## 5. First look at the data

Before running detailed validation checks, it is useful to see a few records.

This helps us understand what one customer record looks like and whether the columns make sense at a basic level.

In [15]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## 6. Check the size of the dataset

The shape tells us:

- number of rows = number of customer records
- number of columns = number of variables available for analysis

This is one of the first things we should establish when receiving a new dataset.

In [16]:
number_of_rows = df.shape[0]
number_of_columns = df.shape[1]

print(f"Number of rows: {number_of_rows:,}")
print(f"Number of columns: {number_of_columns:,}")

Number of rows: 7,043
Number of columns: 21


### Finding

The dataset contains **7,043 customer records and 21 columns**.

This gives us a reasonable-sized dataset for a customer churn classification project and later customer segmentation analysis.

## 7. Check all column names

Next, we verify which variables are actually present in the dataset.

This is important because our later code will depend on these column names.

In [17]:
print("Columns available in the dataset:")
print()

for column_number, column_name in enumerate(df.columns, start=1):
    print(f"{column_number:02d}. {column_name}")

Columns available in the dataset:

01. customerID
02. gender
03. SeniorCitizen
04. Partner
05. Dependents
06. tenure
07. PhoneService
08. MultipleLines
09. InternetService
10. OnlineSecurity
11. OnlineBackup
12. DeviceProtection
13. TechSupport
14. StreamingTV
15. StreamingMovies
16. Contract
17. PaperlessBilling
18. PaymentMethod
19. MonthlyCharges
20. TotalCharges
21. Churn


### Expected business areas

From the column names, we can already see several useful groups of information:

- Customer information: `gender`, `SeniorCitizen`, `Partner`, `Dependents`
- Customer relationship: `tenure`, `Contract`
- Services: `PhoneService`, `InternetService`, `OnlineSecurity`, `TechSupport`, etc.
- Billing: `PaymentMethod`, `MonthlyCharges`, `TotalCharges`
- Target: `Churn`

We will not make modelling decisions from column names alone. Those decisions will be validated during EDA and feature engineering.

## 8. Check the DataFrame information

`df.info()` gives us a useful structural overview:

- column names
- number of non-null values
- data types
- memory usage

This is particularly useful for identifying columns whose stored data type does not match their business meaning.

In [18]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   str    
 1   gender            7043 non-null   str    
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   str    
 4   Dependents        7043 non-null   str    
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   str    
 7   MultipleLines     7043 non-null   str    
 8   InternetService   7043 non-null   str    
 9   OnlineSecurity    7043 non-null   str    
 10  OnlineBackup      7043 non-null   str    
 11  DeviceProtection  7043 non-null   str    
 12  TechSupport       7043 non-null   str    
 13  StreamingTV       7043 non-null   str    
 14  StreamingMovies   7043 non-null   str    
 15  Contract          7043 non-null   str    
 16  PaperlessBilling  7043 non-null   str    
 17  Paymen

### Important observation

One column needs special attention:

`TotalCharges`

It represents a monetary value, but in the raw dataset it is stored as text.

We will investigate this carefully rather than immediately changing it.

## 9. Check for missing values

Missing values can affect analysis, feature engineering, and model training.

First, we check the number of actual null values in every column.

In [19]:
missing_values = df.isnull().sum()

print("Missing values in each column:")
print(missing_values)

Missing values in each column:
customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


### Make the result easier to read

Instead of looking at a long Series, we create a small table containing only columns where null values exist.

If this table is empty, there are no explicit null values.

In [20]:
missing_value_summary = pd.DataFrame({
    "Column": missing_values.index,
    "Missing_Values": missing_values.values
})

missing_value_summary = missing_value_summary[
    missing_value_summary["Missing_Values"] > 0
]

if missing_value_summary.empty:
    print("No explicit missing values were found.")
else:
    display(missing_value_summary)

No explicit missing values were found.


### Finding

The raw dataset reports **0 explicit null values**.

However, this does not completely prove that the dataset has no missing information.

A text column can contain an empty string or whitespace instead of a true null value.

So we perform one more check.

## 10. Check for blank and whitespace values

This check is especially important for text columns.

For example:

`""`

and

`"   "`

are not necessarily detected by `isnull()`.

We therefore check whether any text column contains values that become empty after removing surrounding whitespace.

In [21]:
text_columns = df.select_dtypes(include="object").columns

blank_value_counts = {}

for column in text_columns:
    blank_count = df[column].fillna("").astype(str).str.strip().eq("").sum()
    blank_value_counts[column] = blank_count

blank_value_summary = pd.Series(
    blank_value_counts,
    name="Blank_or_Whitespace_Values"
)

blank_value_summary = blank_value_summary[
    blank_value_summary > 0
]

if blank_value_summary.empty:
    print("No blank or whitespace-only values were found in text columns.")
else:
    display(blank_value_summary)

C:\Users\Aman\AppData\Local\Temp\ipykernel_20876\1621702165.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  text_columns = df.select_dtypes(include="object").columns


TotalCharges    11
Name: Blank_or_Whitespace_Values, dtype: int64

## 11. Check for duplicate rows

Duplicate records can distort customer counts and downstream analysis.

We first check for exact duplicate rows, where every value across the row is identical.

In [22]:
duplicate_rows = df.duplicated().sum()

print(f"Number of exact duplicate rows: {duplicate_rows:,}")

Number of exact duplicate rows: 0


In [23]:
if duplicate_rows == 0:
    print("No exact duplicate rows were found.")
else:
    print("Duplicate rows were found and need investigation before modelling.")

No exact duplicate rows were found.


### Finding

The dataset contains **0 exact duplicate rows**.

Therefore, there is no evidence of exact row duplication at this stage.

We will still separately check whether the customer identifier is unique.

## 12. Validate the customer identifier

`customerID` is an identifier.

An identifier is useful for tracing a prediction back to a customer, but it should generally not be used as a predictive feature because the ID itself does not represent customer behaviour.

Before making that decision, we check whether each customer ID is unique.

In [24]:
customer_id_column = "customerID"

number_of_unique_customers = df[customer_id_column].nunique()
number_of_customer_records = len(df)

print(f"Total customer records: {number_of_customer_records:,}")
print(f"Unique customer IDs: {number_of_unique_customers:,}")

Total customer records: 7,043
Unique customer IDs: 7,043


In [25]:
duplicate_customer_ids = df[customer_id_column].duplicated().sum()

print(f"Duplicate customer IDs: {duplicate_customer_ids:,}")

Duplicate customer IDs: 0


In [26]:
missing_customer_ids = df[customer_id_column].isnull().sum()

print(f"Missing customer IDs: {missing_customer_ids:,}")

Missing customer IDs: 0


### Decision

If every row has one unique customer ID and there are no missing IDs, we can use `customerID` for traceability while excluding it from the predictive feature set later.

We are **not dropping it from the raw dataset**. That decision belongs to the modelling pipeline.

## 13. Validate expected numeric columns

A column should have a data type that makes sense for its business meaning.

For this dataset, we expect these fields to be numerical:

- `SeniorCitizen`
- `tenure`
- `MonthlyCharges`
- `TotalCharges`

Let's check their actual types.

In [27]:
expected_numeric_columns = [
    "SeniorCitizen",
    "tenure",
    "MonthlyCharges",
    "TotalCharges"
]

for column in expected_numeric_columns:
    print(f"{column:20} -> {df[column].dtype}")

SeniorCitizen        -> int64
tenure               -> int64
MonthlyCharges       -> float64
TotalCharges         -> str


### Finding

`TotalCharges` is stored as a string rather than a numeric type.

This does not automatically mean that the values are wrong. It means we need to investigate the actual contents before cleaning the column.

## 14. Investigate `TotalCharges`

`TotalCharges` should represent the total amount charged to a customer.

We will test which values can be converted to numeric values.

We use `errors="coerce"` only for this validation test. We are **not changing the original DataFrame** here.

In [28]:
total_charges_as_numeric = pd.to_numeric(
    df["TotalCharges"].astype(str).str.strip(),
    errors="coerce"
)

conversion_failures = total_charges_as_numeric.isnull().sum()

print(
    f"Number of TotalCharges values that cannot be converted to numeric: "
    f"{conversion_failures:,}"
)

Number of TotalCharges values that cannot be converted to numeric: 11


If conversion failures exist, we need to see the original values responsible for them.

In [29]:
if conversion_failures > 0:

    problematic_total_charges = df.loc[
        total_charges_as_numeric.isnull(),
        "TotalCharges"
    ]

    print("Problematic raw TotalCharges values:")
    display(problematic_total_charges.value_counts(dropna=False))

else:
    print("All TotalCharges values can be converted to numeric.")

Problematic raw TotalCharges values:


TotalCharges
    11
Name: count, dtype: int64

### Validation decision

We have now identified the problem without modifying the raw dataset.

The actual cleaning rule for `TotalCharges` will be decided in the **Data Cleaning** stage, where we can also compare the affected records with `tenure` and other relevant fields.

This keeps validation and cleaning separate.

## 15. Validate categorical columns

Categorical variables should contain consistent values.

For example, if a column is expected to contain:

`Yes`, `No`

we should not silently have values such as:

`yes`, `YES`, `Y`, `No `

We inspect the unique values and their frequencies before deciding whether any standardization is required.

In [30]:
categorical_columns = df.select_dtypes(include="object").columns

print("Categorical columns:")
for column in categorical_columns:
    print(f"- {column}")

Categorical columns:
- customerID
- gender
- Partner
- Dependents
- PhoneService
- MultipleLines
- InternetService
- OnlineSecurity
- OnlineBackup
- DeviceProtection
- TechSupport
- StreamingTV
- StreamingMovies
- Contract
- PaperlessBilling
- PaymentMethod
- TotalCharges
- Churn


C:\Users\Aman\AppData\Local\Temp\ipykernel_20876\481494213.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_columns = df.select_dtypes(include="object").columns


### Inspect category frequencies

We will inspect the categories column by column.

This is intentionally done as a validation step. We are not changing category labels yet.

In [31]:
for column in categorical_columns:

    print()
    print("=" * 70)
    print(f"Column: {column}")
    print("=" * 70)

    value_counts = df[column].value_counts(dropna=False)

    display(value_counts.to_frame(name="Count"))


Column: customerID


,Count
customerID,
7590-VHVEG,1
5575-GNVDE,1
3668-QPYBK,1
7795-CFOCW,1
9237-HQITU,1
...,...
6840-RESVB,1
2234-XADUH,1
4801-JZAZL,1



Column: gender


,Count
gender,
Male,3555
Female,3488



Column: Partner


,Count
Partner,
No,3641
Yes,3402



Column: Dependents


,Count
Dependents,
No,4933
Yes,2110



Column: PhoneService


,Count
PhoneService,
Yes,6361
No,682



Column: MultipleLines


,Count
MultipleLines,
No,3390
Yes,2971
No phone service,682



Column: InternetService


,Count
InternetService,
Fiber optic,3096
DSL,2421
No,1526



Column: OnlineSecurity


,Count
OnlineSecurity,
No,3498
Yes,2019
No internet service,1526



Column: OnlineBackup


,Count
OnlineBackup,
No,3088
Yes,2429
No internet service,1526



Column: DeviceProtection


,Count
DeviceProtection,
No,3095
Yes,2422
No internet service,1526



Column: TechSupport


,Count
TechSupport,
No,3473
Yes,2044
No internet service,1526



Column: StreamingTV


,Count
StreamingTV,
No,2810
Yes,2707
No internet service,1526



Column: StreamingMovies


,Count
StreamingMovies,
No,2785
Yes,2732
No internet service,1526



Column: Contract


,Count
Contract,
Month-to-month,3875
Two year,1695
One year,1473



Column: PaperlessBilling


,Count
PaperlessBilling,
Yes,4171
No,2872



Column: PaymentMethod


,Count
PaymentMethod,
Electronic check,2365
Mailed check,1612
Bank transfer (automatic),1544
Credit card (automatic),1522



Column: TotalCharges


,Count
TotalCharges,
20.2,11
,11
19.75,9
19.9,8
20.05,8
...,...
1990.5,1
7362.9,1
346.45,1



Column: Churn


,Count
Churn,
No,5174
Yes,1869


### What we are looking for

During this review, pay attention to:

- unexpected spelling variations
- leading or trailing spaces
- unexpected null-like values
- categories that do not make business sense
- inconsistent representations of the same concept

The goal is not to remove categories just because they are uncommon.

## 16. Basic numerical sanity checks

Now we validate whether numerical values fall into sensible ranges.

This is different from outlier analysis.

For example, a very high monthly charge might be a legitimate customer. But a negative monthly charge would require investigation.

We therefore check obvious business constraints first.

In [32]:
print("Summary of numerical variables:")
df[["SeniorCitizen", "tenure", "MonthlyCharges"]].describe()

Summary of numerical variables:


,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


### Check `SeniorCitizen`

This dataset uses:

- `0` = not a senior citizen
- `1` = senior citizen

Therefore, values outside 0 and 1 would indicate a data-quality issue.

In [33]:
invalid_senior_citizen_values = df[
    ~df["SeniorCitizen"].isin([0, 1])
]

print(
    f"Number of records with invalid SeniorCitizen values: "
    f"{len(invalid_senior_citizen_values):,}"
)

Number of records with invalid SeniorCitizen values: 0


### Check `tenure`

Customer tenure represents the number of months the customer has stayed with the company.

Negative tenure would not make business sense.

In [34]:
negative_tenure_records = df[df["tenure"] < 0]

print(
    f"Number of records with negative tenure: "
    f"{len(negative_tenure_records):,}"
)

Number of records with negative tenure: 0


### Check `MonthlyCharges`

Monthly charges should not be negative.

A value of zero may or may not require business investigation, but a negative charge is clearly suspicious.

In [35]:
negative_monthly_charge_records = df[
    df["MonthlyCharges"] < 0
]

print(
    f"Number of records with negative MonthlyCharges: "
    f"{len(negative_monthly_charge_records):,}"
)

Number of records with negative MonthlyCharges: 0


## 17. Validate the target variable

The target variable for our classification problem is:

`Churn`

Before modelling, we need to confirm:

1. Which labels exist?
2. Are there unexpected labels?
3. How many customers belong to each class?

In [36]:
target_column = "Churn"

print("Unique target values:")
print(df[target_column].unique())

Unique target values:
<StringArray>
['No', 'Yes']
Length: 2, dtype: str


In [37]:
target_counts = df[target_column].value_counts(dropna=False)

print("Number of customers in each churn class:")
print(target_counts)

Number of customers in each churn class:
Churn
No     5174
Yes    1869
Name: count, dtype: int64


In [38]:
target_percentages = (
    df[target_column]
    .value_counts(normalize=True, dropna=False)
    .mul(100)
    .round(2)
)

print("Percentage of customers in each churn class:")
print(target_percentages)

Percentage of customers in each churn class:
Churn
No     73.46
Yes    26.54
Name: proportion, dtype: float64


### Finding

The dataset contains:

- **5,174 customers who did not churn**
- **1,869 customers who churned**

That means approximately:

- **73.46% No Churn**
- **26.54% Churn**

This is not a reason to remove or artificially rebalance the dataset at this stage.

However, it tells us something important for modelling:

> Accuracy alone will not be enough to evaluate our churn model.

Later we will consider metrics such as recall, precision, F1-score and probability/threshold behaviour.

## 18. Check for obvious target problems

We expect the target to contain only:

- `Yes`
- `No`

Let's explicitly validate that assumption.

In [39]:
expected_churn_values = {"Yes", "No"}

actual_churn_values = set(
    df[target_column].dropna().unique()
)

unexpected_churn_values = actual_churn_values - expected_churn_values

if len(unexpected_churn_values) == 0:
    print("Target validation passed: only Yes and No labels are present.")
else:
    print("Unexpected target labels were found:")
    print(unexpected_churn_values)

Target validation passed: only Yes and No labels are present.


## 19. Basic leakage screening

Data leakage happens when a model receives information that would not realistically be available when the prediction is made.

At this stage we only perform an initial screening.

We will make final feature decisions later after understanding the business problem and the prediction timing.

In [40]:
print("Columns that require timing/business review before modelling:")
print()

columns_to_review = [
    "tenure",
    "TotalCharges",
    "Contract",
    "PaymentMethod"
]

for column in columns_to_review:
    print(f"- {column}")

Columns that require timing/business review before modelling:

- tenure
- TotalCharges
- Contract
- PaymentMethod


### Important note

A column is not automatically leakage just because it is strongly related to churn.

The real question is:

> **Would this information be available at the exact moment when the business wants to make the churn prediction?**

We will revisit this during feature engineering and modelling.

## 20. Create a validation summary

At the end of the notebook, we want a concise summary that another person can review quickly.

This summary does not replace the detailed checks above. It gives us a final project checkpoint.

In [41]:
total_null_values = df.isnull().sum().sum()

total_blank_values = blank_value_summary.sum()

validation_results = {
    "Dataset loaded": True,
    "Exact duplicate rows": duplicate_rows == 0,
    "Duplicate customer IDs": duplicate_customer_ids == 0,
    "Missing customer IDs": missing_customer_ids == 0,
    "No explicit null values": total_null_values == 0,
    "No blank text values": total_blank_values == 0,
    "TotalCharges requires review": conversion_failures > 0,
    "No invalid SeniorCitizen values": len(invalid_senior_citizen_values) == 0,
    "No negative tenure values": len(negative_tenure_records) == 0,
    "No negative MonthlyCharges": len(negative_monthly_charge_records) == 0,
    "Target labels are valid": len(unexpected_churn_values) == 0
}

validation_summary = pd.DataFrame(
    {
        "Validation Check": validation_results.keys(),
        "Passed": validation_results.values()
    }
)

validation_summary

,Validation Check,Passed
0,Dataset loaded,True
1,Exact duplicate rows,True
2,Duplicate customer IDs,True
3,Missing customer IDs,True
4,No explicit null values,True
5,No blank text values,False
6,TotalCharges requires review,True
7,No invalid SeniorCitizen values,True
8,No negative tenure values,True
9,No negative MonthlyCharges,True


## 21. Final validation conclusion

### What we established

The raw dataset contains **7,043 customer records and 21 columns**.

The validation stage has established the following:

- The dataset can be loaded successfully.
- The expected customer-level structure is present.
- There are no exact duplicate rows.
- The customer identifier can be evaluated separately from predictive features.
- There are no explicit null values reported by Pandas.
- Text fields were also checked for blank/whitespace-only values.
- `TotalCharges` requires special attention because it is stored as text.
- Numerical variables were checked for obvious invalid values.
- The target contains `Yes` and `No` churn labels.
- Churn represents approximately 26.54% of the dataset.
- Accuracy alone should not be our primary modelling metric.
- Potential timing/leakage considerations have been identified for later review.

### What we are NOT doing yet

We are deliberately not doing the following in this notebook:

- filling missing values
- converting `TotalCharges` permanently
- removing columns
- encoding categorical variables
- scaling numerical variables
- removing outliers
- selecting features
- training models

Those decisions belong to later stages.

### Next step

**Next notebook: `02_eda_business_analysis.ipynb`**

There we will use the validated dataset to answer actual business questions:

- Which customer groups churn more?
- How does tenure relate to churn?
- How does contract type relate to churn?
- Which services are associated with higher/lower churn?
- Does payment method show a meaningful pattern?
- Which findings are strong enough to influence feature engineering?

> **The purpose of EDA is not to make as many charts as possible. The purpose is to discover evidence that can improve our business understanding and modelling decisions.**